core/ledger/kvledger/unjoin_channel.go

```diff
@@ -11,15 +11,58 @@ import (
 	"github.com/hyperledger/fabric-protos-go/ledger/rwset/kvrwset"
 	"github.com/hyperledger/fabric-protos-go/peer"
 	"github.com/hyperledger/fabric/common/ledger/blkstorage"
+	"github.com/hyperledger/fabric/common/ledger/util/leveldbhelper"
 	"github.com/hyperledger/fabric/common/metrics/disabled"
 	"github.com/hyperledger/fabric/core/ledger"
 	"github.com/hyperledger/fabric/core/ledger/confighistory"
 	"github.com/hyperledger/fabric/core/ledger/kvledger/bookkeeping"
 	"github.com/hyperledger/fabric/core/ledger/kvledger/history"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/msgs"
 	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/privacyenabledstate"
 	"github.com/hyperledger/fabric/core/ledger/pvtdatastorage"
+	"github.com/pkg/errors"
 )
 
+// UnjoinChannel removes the data for a ledger and sets the status to UNDER_DELETION.  This function is to be
+// invoked while the peer is shut down.
+func UnjoinChannel(config *ledger.Config, ledgerID string) error {
+	// Ensure the routine is invoked while the peer is down.
+	fileLock := leveldbhelper.NewFileLock(fileLockPath(config.RootFSPath))
+	if err := fileLock.Lock(); err != nil {
+		return errors.Wrap(err, "as another peer node command is executing,"+
+			" wait for that command to complete its execution or terminate it before retrying")
+	}
+	defer fileLock.Unlock()
+
+	// Set the ledger to a pending deletion status.  At the next startup, the peer will complete the removal of any partially deleted ledgers.
+	if err := updateLedgerStatus(config, ledgerID, msgs.Status_UNDER_DELETION); err != nil {
+		return errors.Wrapf(err, "Unjoin channel [%v]", ledgerID)
+	}
+
+	// remove the ledger data
+	if err := removeLedgerData(config, ledgerID); err != nil {
+		return errors.Wrapf(err, "deleting ledger [%v]", ledgerID)
+	}
+
+	logger.Infof("The channel [%s] has been successfully unjoined", ledgerID)
+	return nil
+}
+
+// Update a ledger status
+func updateLedgerStatus(config *ledger.Config, ledgerID string, status msgs.Status) error {
+	idStore, err := openIDStore(LedgerProviderPath(config.RootFSPath))
+	if err != nil {
+		return err
+	}
+	defer idStore.db.Close()
+
+	if err := idStore.updateLedgerStatus(ledgerID, status); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
Hm... I mildly disagree with this pattern / convention.  It's OK but I actually prefer the explicit assignment and test of the error routine for absolute consistency.  It seems like many (all?) golang routines consist of a serial flow of do-something+check-error+do-something+check-error+... invocations.    When then final invocation in a series of calls is used as the error for the outer function, the reader needs to stop and think about the return type of the last call in the chain, rather than recognize the pattern as a a cookie cutter linear flow.  For instance, in the two examples above, the explicit 'return nil' case is 100% clear that the routine is returning an OK result.  In the latter, looking at the bottom of the routine it's not 100% clear that the method is returning an error, the new status, or the previous status of the ledger.   Call me simple but I prefer the explicit style - the intent is clearer, and leaves opportunities for post-call logging, future method calls without restructuring, and ... ease of use when attached to a debugger.  IMHO it's worth the extra typing. :)